### DISCUSSION TITLE
Not possible to read multisampled depth texture's resolved output

### DISCUSSION THREAD

**Discussion by ottoville** [gpuweb/gpuweb#2380] [Q&A]
With current spec, it is possible to copy depth texture to buffer with "copyTextureToBuffer" and read it into the CPU memory.
However, when MSAA is used, depth texture must have sample count of 4 and "copyTextureToBuffer" does not allow usages with multisampled textures.
While it is possible to read any other multisampled render target into CPU by reading the texture specified in "resolve target" property of GPURenderPassColorAttachment, this is not possible with depth texture, because "resolve target" for depth texture is not configurable in GPURenderPassDepthStencilAttachment.

**Comment by Kangz** [comment]
If you're ok with reading the resolved texture, then you can do a `copyTextureToBuffer` of the resolved texture, which is possible for some of the depth formats. Otherwise you can also use a compute shader or a fragment shader that loads from the multisampled texture directly and puts the values in a storage buffer using this overload:

`textureLoad(t: texture_depth_multisampled_2d, coords: vec2<i32>, sample_index: i32)-> f32`

**Reply by ottoville** [reply]
"If you're ok with reading the resolved texture"

The problem is  that current API does not give any access to resolved texture. It is not configurable in RenderPass like other multisampled render targets are.

**Comment by kvark** [comment]
> While it is possible to read any other multisampled render target into CPU by reading the texture specified in "resolve target" property of GPURenderPassColorAttachment

I think this is the main misunderstanding. You aren't reading the multisampled target, you are reading the result of the MSAA resolve.
Reading the *actual* multisampled target is exactly as difficult for color as it is for depth - you have to read it from the shader by `textureLoad(..., sample_index)`.

**Reply by kainino0x** [reply]
I think this is the correct answer.

To be more specific, the overload you need is
`textureLoad(t: texture_depth_multisampled_2d, coords: vec2<i32>, sample_index: i32)-> f32`
(EDIT: just saw kangz already posted this overload above)

**Comment by Kangz** [comment]
Ah right, yeah you need to sample in a shader.
